README.md
# nos-codamos

Onboarding com os desafios da Alura.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-07-29
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-07-29
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/nos-codamos/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/nos-codamos/compare/0.1.0...0.1.1


test/nos_codamos/core_test.clj
(ns nos-codamos.core-test
  (:require [clojure.test :refer :all]
            [nos-codamos.core :refer :all]))



test/nos_codamos/main_test.clj
(ns nos-codamos.main-test
  (:require [clojure.test :refer :all]
            [nos-codamos.semana-1.main :refer :all])
  (:import (java.time LocalDate YearMonth)
           (java.time.temporal ChronoUnit)))

(def credit-card-test {:number          "111"
                       :cvv             "222"
                       :expiration-date (YearMonth/parse "2029-09")
                       :limit           3000})

(def purchase-1-test
  (purchase (LocalDate/parse "2021-10-07")
            100
            "iFood"
            "Restaurant"
            (update credit-card-test :limit - 100)))
(def purchase-2-test
  (purchase (LocalDate/parse "2021-10-07")
            200
            "iFood"
            "Restaurant"
            (update credit-card-test :limit - 200)))
(def purchase-3-test
  (purchase (LocalDate/parse "2021-10-07")
            300
            "iFood"
            "Restaurant"
            (update credit-card-test :limit - 300)))
(def purchase-4-test
  (purchase (LocalDate/parse "2021-10-07")
            6000
            "iFood"
            "Restaurant"
            (:credit-card purchase-3-test)))

(def all-purchases-test [purchase-1-test purchase-2-test purchase-3-test purchase-4-test])

(def approved-purchases-test (search-purchases all-purchases-test {:approved? true}))

(deftest valid-amount?-test
  (testing "valid purchase amount")
  (is (not (valid-amount? nil)))
  (is (not (valid-amount? 0)))
  (is (valid-amount? 0.1))
  (is (valid-amount? 1))
  (is (not (valid-amount? -0.1)))
  (is (not (valid-amount? "0.1"))))

(deftest purchase-test
  (testing "adding new purchase")
  (is (= {:date        (LocalDate/parse "2021-05-07")
          :amount      0
          :merchant    "iFood"
          :category    "Restaurant"
          :approved?   false
          :credit-card credit-card-test}
         (purchase (LocalDate/parse "2021-05-07")
                   0
                   "iFood"
                   "Restaurant"
                   credit-card-test)))
  (is (= {:date        (LocalDate/parse "2021-05-07")
          :amount      -1
          :merchant    "iFood"
          :category    "Restaurant"
          :approved?   false
          :credit-card credit-card-test}
         (purchase (LocalDate/parse "2021-05-07")
                   -1
                   "iFood"
                   "Restaurant"
                   credit-card-test)))
  (is (= {:date        (LocalDate/parse "2021-05-07")
          :amount      1
          :merchant    "iFood"
          :category    "Restaurant"
          :approved?   true
          :credit-card (update credit-card-test :limit - 1)}
         (purchase (LocalDate/parse "2021-05-07")
                   1
                   "iFood"
                   "Restaurant"
                   credit-card-test)))
  (is (= {:date        (LocalDate/parse "2021-05-07")
          :amount      5000
          :merchant    "iFood"
          :category    "Restaurant"
          :approved?   false
          :credit-card credit-card-test}
         (purchase (LocalDate/parse "2021-05-07")
                   5000
                   "iFood"
                   "Restaurant"
                   credit-card-test))))

(deftest search-purchases-test
  (testing "listing all approved purchases")
  (is (= (filter #(= (:approved? %) true) all-purchases-test)
         (search-purchases all-purchases-test {:approved? true}))))

(deftest group-purchases-by-category-test
  (testing "listing purchases by category")
  (is (= (list {"Restaurant" 600})
         (group-purchases-by-category approved-purchases-test))))

src/nos_codamos/core.clj
(ns nos-codamos.core
  (:require [nos-codamos.semana-1.main :as semana-1]))

(println "Semana 1")
(println "-------------------------------------------------")
(println "Last transaction:" (last semana-1/all-purchases))
(println "Credit Card most updated:" (:credit-card (last semana-1/all-purchases)))
(println "-------------------------------------------------")
(println "Expenses by category:" (semana-1/group-purchases-by-category semana-1/approved-purchases))
(println "-------------------------------------------------")
(println "Search by merchant using query: {:merchant iFood}:" (semana-1/search-purchases semana-1/all-purchases {:merchant "iFood"}))
(println "Search by amount using query: {:amount 100}:" (semana-1/search-purchases semana-1/all-purchases {:amount 100}))
(println "---------------------------------------------------")
(println "Monthly bill of month 05" (semana-1/monthly-bill semana-1/approved-purchases 2021 5))
(println "Monthly bill of month 10" (semana-1/monthly-bill semana-1/approved-purchases 2021 10))


src/nos_codamos/semana_1/main.clj
(ns nos-codamos.semana-1.main
  (:import (java.time LocalDate YearMonth)
           (java.time.temporal ChronoUnit)))

(def client-data {:full-name "Tiago Vidal"
                  :cpf       "999.999.999-99"
                  :email     "tiago.vidal@my-email.com.br"})

(def credit-card {:number          "111"
                  :cvv             "222"
                  :expiration-date (YearMonth/parse "2029-09")
                  :limit           100})

(defn expired-card?
  [expiration-date purchase-date]
  (let [purchase-date-year-and-month (YearMonth/from purchase-date)
        chrono-months (ChronoUnit/MONTHS)
        diff-in-months (.until purchase-date-year-and-month expiration-date chrono-months)]
    (<= diff-in-months 0)))

(defn limit?
  [limit amount]
  (>= limit amount))

(defn valid-amount?
  [amount]
  (and (number? amount) (pos? amount)))

(defn valid-purchase?
  [limit? expired-card? valid-amount?]
  (and limit? valid-amount? (not expired-card?)))

(defn new-limit
  [limit amount]
  (- limit amount))

(defn purchase
  [date amount merchant category credit-card]
  (let [current-limit (:limit credit-card)
        expiration-date (:expiration-date credit-card)
        valid-purchase (valid-purchase?
                         (limit? current-limit amount)
                         (expired-card? expiration-date date)
                         (valid-amount? amount))
        limit-to-update (if valid-purchase
                          (new-limit current-limit amount)
                          current-limit)
        credit-card-updated (update credit-card :limit (constantly limit-to-update))]
    {:date        date
     :amount      amount
     :merchant    merchant
     :category    category
     :approved?   valid-purchase
     :credit-card credit-card-updated}))

(defn purchases-amount
  [purchases]
  (map :amount purchases))

(defn total-purchases-amount
  [purchases]
  (reduce + (purchases-amount purchases)))

(defn total-purchases-amount-by-category
  [[category-name purchases]]
  {category-name (total-purchases-amount purchases)})

(defn group-purchases-by-category
  [purchases]
  (map
    total-purchases-amount-by-category
    (group-by :category purchases)))

(defn search-purchases
  [purchases query]
  (filter
    (fn [purchase]
      (= query
         (select-keys purchase (keys query))))
    purchases))

(defn search-purchases-by-year-and-month
  [purchases year month]
  (filter
    (fn [purchase]
      (and
        (= month (.getMonthValue (:date purchase)))
        (= year (.getYear (:date purchase)))))
    purchases))

(defn monthly-bill
  [purchases year month]
  (let [monthly-purchases (search-purchases-by-year-and-month purchases year month)]
    (total-purchases-amount monthly-purchases)))

(def purchase-1 (purchase (LocalDate/parse "2021-05-07")
                          0
                          "iFood"
                          "Restaurant"
                          credit-card))

(def purchase-2 (purchase (LocalDate/parse "2021-10-07")
                          95
                          "C&A"
                          "Clothing"
                          (:credit-card purchase-1)))

(def purchase-3 (purchase (LocalDate/parse "2021-10-07")
                          100
                          "C&A"
                          "Clothing"
                          (:credit-card purchase-2)))

(def all-purchases [purchase-1 purchase-2 purchase-3])

(def approved-purchases (search-purchases all-purchases {:approved? true}))